docs/FoLiA-2text.1
.TH FoLiA-2text 1 "2021 jan 19"

.SH NAME
FoLiA-2text - extract text from FoLiA

.SH SYNOPSIS
FoLiA-2text [options] FILES

FoLiA-2text [options] DIR

.SH DESCRIPTION

When a DIR is provided,
.B FoLiA-2text
will process all files in DIR and store the result in DIR, or in
the directory provided with
.B -o

When a FILE is provided,
.B FoLiA-2text
will process that file and store its result in the directory where FILE is
found or in the directory provided with
.B -o

All files should be in FoLiA XML format or BZIP2 or GZ versions thereof.


.SH OPTIONS
.B -o
prefix
.RS
use 'prefix' as a prefix for the output file(s)
.RE

.B -e
expression
.RS
use 'expression' as a value to match all filenames to for the search in the
input DIR
.RE


.B --class
classname
.RS
The FoLiA classname of the text nodes searched. (default 'current').
.RE

.B --retaintok
.RS
Keep the tokenization as stored in the FoLiA. Default is to remove it.
.RE

.B -t
or
.B --threads
number
.RS
Number of concurrent threads to be used by the program.
.RE

.B -v
.RS
be more verbose.
.RE

.B -V
or
.B --version
.RS
Show VERSION
.RE

.B -h
or
.B --help
.RS
Show some help
.RE

.SH BUGS
possible

.SH AUTHORS
Ko van der Sloot

Martin Reynaert

e\-mail: lamasoftware@science.ru.nl


NEWS
foliautils 0.16  07-01-2021
[Ko vd Sloot]
* requires libfolia 2.7 or above
* provenance data is better for a lot of modules
* added better checking on invalid NCnames in some modules.
* FoLiA-abby:
  - a lot of refactoring and additions to handle font/style information
* FoLiA-pm:
  - Notes are handled correctly now
  - fixed error in xlink attributes
* FoLiA-page:
  - more types of Page files are handled now
  - fixed annotation declarations
  - fixed offset calculation (due to change in FoLiA's opinion on those)
  - page number is added as a <br> node and in the metadata
  - added a --trusttokens option. This means that Word items in the Page file
    are added as Word's in the FoLiA, embedded in Sentences.
  - added a --norefs option to avoid adding references to the original texts
* FoLiA-correct:
  - make sure that the default is to run on 1 thread
  - added a --rebase-inputclass option

* FoLiA-alto:
  - the -t option was not always handled correctly

[Maarten van Gompel]
* FoLiA-benchmark: guard against compiler optimisation #48

foliautils 0.15  15-09-2020
[Maarten van Gompel]
* FoLiA-txt: check if a string is empty after normalisation (fix for #46)
[Ko vd Sloot]
* folia-correct: fix one-off error in hemp handling (when no hemp was found) #45
* some refactoring
  - centralized definition of XML_PARSER_OPTIONS
* bugfix in threading

foliautils 0.14  15-04-2020
[Martin Reynaert]
* updated man pages

[Ko vd Sloot]
* added man pages
* revised usage() in many modules
* the default seperator in FoLiA-stats is '_' now
* fix for: https://github.com/LanguageMachines/foliautils/issues/37
* fix for: https://github.com/LanguageMachines/foliautils/issues/41
* adapted to changes in libfolia
* many small code refactorings
* FoLiA-correct is improved a lot, allowing ngram corrections in FoLiA
* FoLiA-stats accepts a 'word_in_doc' mode now
* FoLiA-alto by default ceated <w> nodes now. use --oldstring to get <str>
* improved a lot in tests/
* many small fixes

foliautils 0.13  30-08-2019
[Ko vd Sloot]
bug fixes:
* fix for https://github.com/LanguageMachines/foliautils/issues/35
* fix for https://github.com/LanguageMachines/foliautils/issues/36

[Maarten van Gompel]
new features:
* FoLiA-wordtranslate.cxx: frog should be able to deal with
  spaces now, no need for ugly non-breaking space hack that now causes
  other problems in frog (LanguageMachines/frog#77)
* FoLiA-wordtranslate.cxx: adding ability to constrain by language


foliautils 0.12  29-05-2019
[Ko vd Sloot]
Release for FoLiA 2.0.

foliautils 0.11  15-05-2019
[Ko vd Sloot]
Stabilizing release for pre FoLiA 2.0

* Updated and added some tests
* started moving common code to a separate file and build a library
  (libfoliautils)
  - hemp detection is one of them

* FoLiA-stats:
  - added possibility to read a list of directories + file-names to process
    into seperate output directories. (could be generalized to other programs)
  - better hemp detection
* FoLiA-correct:
  - use same hempo detections as FoLiA-stats
* FoLiA-abby:
  - support more flavors
* FoLiA-clean:
  - avoid removing the last remaining tekt on <w> nodes
  - cleaning of tokenization now works


foliautils 0.10  29-11-2018
[Ko vd Sloot]
  * fixed icu:namespace issues
  * added FoLiA-abby, an ABBY to FoLiA convertor
  * src/FoLiA-abby.cxx, src/FoLiA-page.cxx, src/FoLiA-pm.cxx:
     - Allow 'none' value for --prefix
  * src/FoLiA-page.cxx, src/FoLiA-hocr.cxx: fixed Alignment info
  * src/FoLiA-correct.cxx:
     - fixed a problem with correction of the last word of a trigram.
     - fix correction of paragraphs with only deeper text
     - The --rank option accepts more flavors of files
  * src/FoLiA-stats.cxx:
     - added a --detokenize option
  * several minor fixes, refactorings etc.
  * updated tests

foliautils 0.9.2  05-06-2018
[Ko vd Sloot]
Bug fix release:
 * append small prefixes to output filenames, to ALWAYS avoid names starting with
   a numeric value.
   'FPM-' for FoLiA-pm. 'FP-' for FoLiA-page, 'FH-' for FoLiA-hocr
   Can bet set witth --prefix
 * FoLiA-stats.cxx:
   - added --collect to usage() and 'man' page
 * FoLiA-correct:
   - added --inputclass and --outputclass parameters (must be different)
   - Don't crash on empty text.

foliautils 0.9.1  17-05-2018
[Ko vd Sloot]
Bug fix release: the tests directory wasn't included in the distribution

foliautils 0.9  16-05-2018
[Ko vd Sloot]
* FoLiA-stats.cxx:
  - added a --collect option, to create files with all n-grams
    together
  - clearer message in FoLiA-stats when no results were found
  - extract text from deeper nodes, if needed
  - fixed out-of-bounds problem
  - now fails when every inputfile fails
* FoLiA-txt:
  - now fails when every inputfile fails
* avoid xml:id's starting with a number. Add "id-" in front.
* added more tests

[Maarten van Gompel]
* added codemeta.json


foliautils 0.8  19-02-2018
[Ko vd Sloot]
* added -R option to FoLiA-collect
* FoLiA-collect now can work in parallel (-t option)
* modernized configuration, whit better Max OSX support (including OpenMP)
* all modules end with an exit code now.
* added more tests to 'make check'
* added output of Type-Token Ratio's (also in degrees)
* several bugfixes.
* code cleanup and refactoring, some speedup too

foliautils 0.7  24-10-2017
[ko vd Sloot]
 * updated and expanded tests
 * fixed offset calculations in FoLiA-hocr, FoLiA-page.cxx
   and FoLiA-alto. We use unicode points now. (needed for folia v1.5 and above)
 * Changed 'modes' in FoLiA-stats, to be a bit more comprehensible
 * fixed problem with metadatatype when 'foreign-data' is present
 * enhanced FoLiA-clean. Still not done...
 * switched to dynamic OMP scheduling in most programs.
   (which process files with probably big differences in processing time)
 * small bugfixes.
 * general cleanup and refactoring

[Maarten van Gompel]
 * Added and improved FoLiA-wordtranslate.cxx

foliautils 0.6  04-04-2017
This is an intermediate release!!

src/FoLiA-2text.cxx
/*
  Copyright (c) 2014 - 2021
  CLST  - Radboud University
  ILK   - Tilburg University

  This file is part of foliautils

  foliautils is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 3 of the License, or
  (at your option) any later version.

  foliautils is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, see <http://www.gnu.org/licenses/>.

  For questions and suggestions, see:
      https://github.com/LanguageMachines/foliautils/issues
  or send mail to:
      lamasoftware (at ) science.ru.nl
*/

#include <string>
#include <vector>
#include <iostream>
#include <fstream>

#include "ticcutils/CommandLine.h"
#include "ticcutils/FileUtils.h"
#include "ticcutils/StringOps.h"
#include "libfolia/folia.h"

#include "config.h"
#ifdef HAVE_OPENMP
#include "omp.h"
#endif

using namespace	std;
using namespace	icu;
using namespace	folia;

void usage( const string& name ){
  cerr << "Usage: " << name << " [options] file/dir" << endl;
  cerr << "\t FoLiA-2text will produce a text from a FoLiA file, " << endl;
  cerr << "\t or a whole directory of FoLiA files " << endl;
  cerr << "\t--class='name'\t use 'name' as the folia class for <t> nodes. (default is 'current')" << endl;
  cerr << "\t--retaintok\t retain tokenization. Default is attempt to remove." << endl;
  cerr << "\t-t 'threads' or\n\t--threads='threads' Number of threads to run on." << endl;
  cerr << "\t\t\t If 'threads' has the value \"max\", the number of threads is set to a" << endl;
  cerr << "\t\t\t reasonable value. (OMP_NUM_TREADS - 2)" << endl;
  cerr << "\t-h or --help\t this message" << endl;
  cerr << "\t-V or --version \t show version " << endl;
  cerr << "\t-e\t\t expr: specify the expression all input files should match with." << endl;
  cerr << "\t-o\t\t name of the output file(s) prefix." << endl;
}

int main( int argc, char *argv[] ){
  TiCC::CL_Options opts( "hVvpe:t:o:", "class:,help,version,retaintok,threads:" );
  try {
    opts.init(argc,argv);
  }
  catch( TiCC::OptionError& e ){
    cerr << e.what() << endl;
    usage(argv[0]);
    exit( EXIT_FAILURE );
  }
  string progname = opts.prog_name();
  if ( opts.empty() ){
    usage( progname );
    exit(EXIT_FAILURE);
  }
  string expression;
  string outputPrefix;
  string value;
  if ( opts.extract('V') || opts.extract("version") ){
    cerr << PACKAGE_STRING << endl;
    exit(EXIT_SUCCESS);
  }
  if ( opts.extract('h') || opts.extract("help") ){
    usage(progname);
    exit(EXIT_SUCCESS);
  }
  if ( opts.extract( 'o', outputPrefix ) ){
    // if ( outputPrefix.empty() ){
    //   cerr << "an output filename prefix is required. (-o option) " << endl;
    //   exit(EXIT_FAILURE);
    // }
  }
  bool retaintok = opts.extract( "retaintok" );
  if ( opts.extract('t', value ) || opts.extract("threads", value ) ){
#ifdef HAVE_OPENMP
    int numThreads = 1;
    if ( TiCC::lowercase(value) == "max" ){
      numThreads = omp_get_max_threads() - 2;
    }
    else if ( !TiCC::stringTo(value,numThreads) ) {
      cerr << "illegal value for -t (" << value << ")" << endl;
      exit( EXIT_FAILURE );
    }
    omp_set_num_threads( numThreads );
#else
    cerr << "-t option does not work, no OpenMP support in your compiler?" << endl;
    exit(EXIT_FAILURE);
#endif
  }
  opts.extract('e', expression );
  string class_name = "current";
  opts.extract( "class", class_name );

  vector<string> fileNames = opts.getMassOpts();
  if ( fileNames.empty() ){
    cerr << "no file or dir specified!" << endl;
    exit(EXIT_FAILURE);
  }

  if ( fileNames.size() == 1 && TiCC::isDir( fileNames[0] ) ){
    fileNames = TiCC::searchFilesMatch( fileNames[0], expression );
  }
  size_t toDo = fileNames.size();
  if ( toDo == 0 ){
    cerr << "no matching files found" << endl;
    exit(EXIT_SUCCESS);
  }

  if ( !outputPrefix.empty() ){
    string::size_type pos = outputPrefix.find( "/" );
    if ( pos != string::npos && pos == outputPrefix.length()-1 ){
      // outputname ends with a /
    }
    else {
      outputPrefix += "/";
    }
    if ( !TiCC::createPath( outputPrefix ) ){
      cerr << "Output to '" << outputPrefix << "' is impossible" << endl;
    }
  }
  else {
    outputPrefix = TiCC::dirname( fileNames[0] ) + "/";
  }

  if ( toDo > 1 ){
    cout << "start processing of " << toDo << " files " << endl;
  }

#pragma omp parallel for shared(fileNames) schedule(dynamic)
  for ( size_t fn=0; fn < fileNames.size(); ++fn ){
    string docName = fileNames[fn];
    Document *d = 0;
    try {
      d = new Document( "file='"+ docName + "'" );
    }
    catch ( exception& e ){
#pragma omp critical
      {
	cerr << "failed to load document '" << docName << "'" << endl;
	cerr << "reason: " << e.what() << endl;
      }
      continue;
    }
    string outname = outputPrefix + TiCC::basename(docName) + ".txt";
    if ( !TiCC::createPath( outname ) ){
#pragma omp critical
      {
	cerr << "Output to '" << outname << "' is impossible" << endl;
      }
    }
    else {
      UnicodeString us;
      try {
	us = d->text( class_name, retaintok);
      }
      catch( ... ){
	cout << "document '" << docName << "' contains no text in class="
	     << class_name << endl;
      }
      if ( !us.isEmpty() ){
	ofstream os( outname );
	os << us << endl;
      }
#pragma omp critical
      {
	cout << "Processed :" << docName << " into " << outname
	     << " still " << --toDo << " files to go." << endl;
      }
    }
    delete d;
  }
  return EXIT_SUCCESS;
}
